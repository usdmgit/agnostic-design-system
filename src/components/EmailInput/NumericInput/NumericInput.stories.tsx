import React, { useState } from 'react';
import { Meta } from '@storybook/react';
import NumericInput from '@/components/NumericInput';
import mdx from './NumericInput.stories.mdx';

export default {
  title: 'Components/Numeric Input',
  component: NumericInput,
  parameters: {
    docs: {
      page: mdx
    }
  },
  argTypes: {
    max: {
      control: {
        type: 'range',
        min: 0,
        max: 10000,
        step: 100
      }
    }
  }
} as Meta;

const Template = args => {
  const [value, setValue] = useState('');

  return (
    <NumericInput
      id='numeric-input'
      {...args}
      value={value}
      onChange={e => setValue(e.target.value)}
    />
  );
};

export const Default = Template.bind({});
Default.args = {
  placeholder: 'Enter number',
  label: "You can only type ',', '.' and numbers"
};

export const WithMax = Template.bind({});
WithMax.args = {
  label: 'The max here is 1000',
  max: 1000
};

export const WithPositive = Template.bind({});
WithPositive.args = {
  label: 'You can not type negative numbers here',
  positive: true
};

export const withRadixAndScale = Template.bind({});
withRadixAndScale.args = {
  label: 'The radix here is , and the scale is 2',
  radix: ',',
  scale: 2
};
