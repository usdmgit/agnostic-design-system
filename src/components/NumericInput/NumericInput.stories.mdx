# Numeric Input Component

## Type: Decorator

The Numeric Input component is based off of the [Input](https://design-system.codelitt.dev/?path=/docs/components-input--default) component.

## Numeric Input Props

Given the nature of a decorator component, we use all the properties of the **Input**. Please refer to link above for more details.<br />
The Numeric Input, however, has four unique properties to it, as listed below:

|   Prop   |  Type   |                                                                                  Description                                                                                   | Default | Mandatory |
| :------: | :-----: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :-----: | :-------: |
|   max    | number  |                                                                    Maximum possible value for the component                                                                    |    -    |   false   |
| positive | boolean | Specifies whether or not the input can accept negative numbers. If props is absent, then it is allowed. If give a value of true, then negative numbers are unable to be typed. |    -    |   false   |
|  radix   | string  |                                             Fractional delimiter. For example: a radix of `','` with a `scale: 2` would be _1,00_                                              |    -    |   false   |
|  scale   | number  |                                                                  Digits after decimal point, `0` for integers                                                                  |    0    |   true    |

## Getting Started

```javascript
import React, { useState } from 'react';
import { NumericInput } from '@codelittinc/agnostic-design-system';

const SimpleNumericInput = () => {
  const [value, setValue] = useState('');

  return (
    <NumericInput
      label='Numeric Input'
      onChange={e => setValue(e.target.value)}
      max={15}
      positive={true}
      radix={','}
      scale={2}
      value={value}
    />
  );
};

export default SimpleNumericInput;
```

## CSS Variables Available

Given the nature of a decorator component, we use all the css variables of the **Input** component. Please refer to link above for more details.
