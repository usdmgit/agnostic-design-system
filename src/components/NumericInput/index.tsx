import React from 'react';
// @TODO: Discover how to disable the no-unused-vars verification for ts interfaces
/* eslint-disable */
import Input, { Filter, Size, Validation, Props as InputProps } from '../Input';
  /* eslint-enable */

interface Props {
  actionIcon?: React.ReactNode;
  description?: string;
  disabled?: boolean;
  filters?: Filter[];
  id: string;
  label?: string | React.ReactNode;
  limit?: number;
  max?: number;
  message?: string;
  onBlur?: () => void;
  onChange: (e: any) => void;
  onClickActionIcon?: () => void;
  onFocus: () => void;
  onKeyDown?: () => void;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
  onStateChange: (state: boolean) => void;
  placeholder?: string;
  prepend?: React.ReactNode;
  positive?: boolean;
  radix?: string;
  required?: boolean;
  scale: number;
  size: Size;
  value?: string;
  validations?: Validation[];
  variablesClassName?: string;
  withActionIcon?: boolean;
  withPrependSeparator?: boolean;
}

const NON_DIGITS_REGEX = new RegExp(/\D+/, 'g');

const isNumberBiggerThanMax = (radix, num, max) => {
  if (radix) {
    const splitNumbers = num.split(radix);
    const removedPlacementsNumber = splitNumbers[0].replace(NON_DIGITS_REGEX, '');
    const decimalValue = splitNumbers[1];
    const result = Number(`${removedPlacementsNumber}.${decimalValue || 0}`);
    return max < result;
  } else {
    const numberWithoutRadix = num.replace(NON_DIGITS_REGEX, '');
    return max < numberWithoutRadix;
  }
};

const NumericInput = React.forwardRef<HTMLInputElement, Props>((props, ref) => {
  const { value, scale, max, radix, positive, ...inputProps } = props;
  const SCALE_REGEX = new RegExp(`^.+${radix}\\d{${scale + 1}}`);

  const filters = [
    {
      type: 'RegExp',
      test: /[^\d.,-]/
    },
    {
      type: 'function',
      test: value => scale && SCALE_REGEX.test(value)
    },
    {
      type: 'function',
      test: value => positive && value.lastIndexOf('-') !== -1
    },
    {
      type: 'function',
      test: value => isNumberBiggerThanMax(radix, value, max)
    }
  ] as InputProps['filters'];

  return <Input {...inputProps} ref={ref} value={value} filters={filters} />;
});

NumericInput.defaultProps = {
  size: 'large',
  onFocus: () => {},
  scale: 0
};

export default NumericInput;
