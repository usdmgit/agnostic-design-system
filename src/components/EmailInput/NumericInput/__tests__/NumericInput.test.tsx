import React, { useState } from 'react';
import NumericInput from '@/components/NumericInput';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const getNumericInputWrapperComponent = (max?: number, radix?: string, scale: number = 2) => {
  const Wrapper = () => {
    const [value, onChange] = useState('');
    return (
      <NumericInput
        id='numeric-input-test'
        onChange={e => onChange(e.target.value)}
        onFocus={jest.fn}
        onStateChange={jest.fn}
        max={max}
        placeholder='Enter Number'
        radix={radix}
        scale={scale}
        size='large'
        value={value}
      />
    );
  };

  return render(<Wrapper />);
};

const getInput = (container: HTMLElement): HTMLInputElement => {
  return container.querySelector<HTMLInputElement>('input')!;
};

describe('NumericInput', () => {
  it('renders correctly', () => {
    const { container } = getNumericInputWrapperComponent();
    expect(container).toMatchSnapshot();
  });

  describe('Default NumericInput', () => {
    it('allows to valid numeric values', () => {
      const { container } = getNumericInputWrapperComponent();
      const input = getInput(container);
      fireEvent.change(input, { target: { value: 123 } });
      expect(input.value).toBe('123');
    });

    it('blocks invalid characters from being passed', () => {
      const { container } = getNumericInputWrapperComponent();
      const input = getInput(container);
      fireEvent.change(input, { target: { value: 'test' } });
      const expectedResult = '';
      expect(input.value).toBe(expectedResult);
    });

    it('prevents invalid characters from being passed alongside valid characters', () => {
      const { container } = getNumericInputWrapperComponent();
      const input = getInput(container);
      userEvent.type(input, '123abc');
      const expectedResult = '123';
      expect(input.value).toBe(expectedResult);
    });
  });

  describe('Max-defined Numeric Input', () => {
    describe('Values equal-to or less than the max', () => {
      it('accepts values equal to the max as valid', () => {
        const { container } = getNumericInputWrapperComponent(1000);
        const input = getInput(container);
        userEvent.type(input, '1000');
        expect(input.value).toBe('1000');
      });

      it('accepts values less than the max as valid', () => {
        const { container } = getNumericInputWrapperComponent(1000);
        const input = getInput(container);
        userEvent.type(input, '100');
        expect(input.value).toBe('100');
      });
    });

    describe('Values exceeding the max', () => {
      it('prevents values exceeding the max from being passed', () => {
        const { container } = getNumericInputWrapperComponent(1000);
        const input = getInput(container);
        userEvent.type(input, '10000');
        const expectedResult = '1000';
        expect(input.value).toBe(expectedResult);
      });

      it('prevents values with radix exceeding the max from being passed', () => {
        const { container } = getNumericInputWrapperComponent(1000, ',');
        const input = getInput(container);
        userEvent.type(input, '1000,99');
        const expectedResult = '1000,';
        expect(input.value).toBe(expectedResult);
      });

      it('prevents values with placements exceeding the max from being passed', () => {
        const { container } = getNumericInputWrapperComponent(5000000);
        const input = getInput(container);
        userEvent.type(input, '12.521.215,');
        const expectedResult = '12.521.21';
        expect(input.value).toBe(expectedResult);
      });
    });
  });

  describe('Radix-defined Numeric Input', () => {
    it('accepts comma-defined radix', () => {
      const { container } = getNumericInputWrapperComponent(100, ',');
      const input = getInput(container);
      userEvent.type(input, '100,00');
      expect(input.value).toBe('100,00');
    });

    it('accepts dot-defined radix', () => {
      const { container } = getNumericInputWrapperComponent(100, '.');
      const input = getInput(container);
      userEvent.type(input, '100.00');
      expect(input.value).toBe('100.00');
    });
  });

  describe('Scale-defined Numeric Input', () => {
    it('accepts scale within the defined limit', () => {
      const { container } = getNumericInputWrapperComponent(undefined, '.', 2);
      const input = getInput(container);
      userEvent.type(input, '100.00');
      expect(input.value).toBe('100.00');
    });

    it('prevents decimal value placements from exceeding the scale defined limit', () => {
      const { container } = getNumericInputWrapperComponent(undefined, ',', 2);
      const input = getInput(container);
      userEvent.type(input, '100,101');
      const expectedResult = '100,10';
      expect(input.value).toBe(expectedResult);
    });
  });
});
