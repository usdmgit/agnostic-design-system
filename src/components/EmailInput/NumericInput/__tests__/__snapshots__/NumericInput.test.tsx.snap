// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericInput renders correctly 1`] = `
<div>
  <div
    class="container"
  >
    
    
    <div
      class="input--field-container"
    >
      <div
        class="input--container"
      >
        <input
          class="input input--large"
          id="numeric-input-test"
          placeholder="Enter Number"
          value=""
        />
      </div>
      
    </div>
  </div>
</div>
`;
